pkg/provisioner/lvm.go

```diff
@@ -0,0 +1,210 @@
+	}
+	requeue, err := l.addDevOrCreateLVMCgCRD(lvmvg, notFound)
+	if err != nil {
+		return requeue, err
+	}
+
+	// first round the lvmvg must be nil, so we need to check it.
+	if lvmvg != nil && lvmvg.Status != nil && lvmvg.Status.Status == diskv1.VGStatusActive {
+		setCondDiskAddedToNodeTrue(l.device, fmt.Sprintf("Added disk %s to volume group %s ", l.device.Name, l.vgName), diskv1.ProvisionPhaseProvisioned)
+		return false, nil
+	}
+	return true, nil
+}
+
+// UnProvision update the LVMVg CRD and remove the LVMVg CRD if the device is the last one in the VG.
+func (l *LVMProvisioner) UnProvision() (bool, error) {
+	// check if the LVMVg CRD is already provisioned
+	targetName := fmt.Sprintf(l.nodeName + "-" + l.vgName)
+	lvmvg, err := l.vgClient.Get(utils.HarvesterNS, targetName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// do nothing if the LVMVg CRD is not found
+			logrus.Warnf("LVMVG %s is not found, skip UnProvision", targetName)
+			msg := fmt.Sprintf("Removed disk %s from volume group %s ", l.device.Name, l.vgName)
+			setCondDiskAddedToNodeFalse(l.device, msg, diskv1.ProvisionPhaseUnprovisioned)
+			return false, nil
+		}
+		return true, fmt.Errorf("failed to get LVMVolumeGroup %s. %v", l.vgName, err)
+	}
+	logrus.Infof("%s unprovisioning block device %s from vg: %s", l.name, l.device.Name, l.vgName)
+	requeue, err := l.removeDevFromLVMVgCRD(lvmvg, l.device.Name)
+	if err != nil {
+		return requeue, err
+	}
+	if lvmvg.Status != nil {
+		if _, found := lvmvg.Status.Devices[l.device.Name]; !found {
+			msg := fmt.Sprintf("Removed disk %s from volume group %s ", l.device.Name, l.vgName)
+			setCondDiskAddedToNodeFalse(l.device, msg, diskv1.ProvisionPhaseUnprovisioned)
+			return false, nil
+		}
+	}
+	// waiting the device removed from the LVMVg CRD
+	logrus.Infof("Waiting for the device %s removed from the LVMVg CRD %v", l.device.Name, lvmvg)
+	return true, nil
+}
+
+func (l *LVMProvisioner) Update() (requeue bool, err error) {
+	// Update DesireState to Reconciling
+	logrus.Infof("Prepare to Update LVMVolumeGroup %s", l.vgName)
+	targetName := fmt.Sprintf(l.nodeName + "-" + l.vgName)
+	lvmvg, err := l.vgClient.Get(utils.HarvesterNS, targetName, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return true, fmt.Errorf("failed to get LVMVolumeGroup %s, err: %v", targetName, err)
+	}
+
+	if lvmvg.Spec.DesireState == diskv1.VGStateEnabled {
+		// make sure the volume group is active
+		err := lvm.DoVGActivate(lvmvg.Spec.VgName)
+		if err != nil {
+			return true, fmt.Errorf("failed to activate volume group %s, err: %v", l.vgName, err)
+		}
+	} else if lvmvg.Spec.DesireState == diskv1.VGStateDisabled {
+		// make sure the volume group is inactive
+		logrus.Infof("Should not go here, because the LVMVolumeGroup %s should not be disabled", l.vgName)
+	}
+	return
+}
+
+func (l *LVMProvisioner) addDevOrCreateLVMCgCRD(lvmVG *diskv1.LVMVolumeGroup, notFound bool) (requeue bool, err error) {
```

Comment:
```suggestion
func (l *LVMProvisioner) addDevOrCreateLVMVgCRD(lvmVG *diskv1.LVMVolumeGroup, notFound bool) (requeue bool, err error) {
```

Nit. might need a rename too.

It would be nice to be consistent. The terms "LVMVolumeGroup", "LVMVG", and "LVMVg" are used in the log messages.

Reply:
Align with `LVMVolumeGroup` for the log messages.